### PR TITLE
Large partition new flow

### PR DESCRIPTION
**Pull request by juliayakovlev**
## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

1. new ScyllaBenchEvent added (with unit test) and add it to the filter on Nemesis Dashboards
2. cause scylla-bench run with round-robin
3. new large partition flow

[Trello task](https://trello.com/c/hDVG6iGx/1305-large-partitions-workload)

### CODE REVIEW

File: unit_tests/test_events.py [hunk 1]
```diff
@@ -10,13 +10,15 @@
...
 from sdcm.sct_events import (start_events_device, stop_events_device, Event, TestKiller,
-                             InfoEvent, CassandraStressEvent, CoreDumpEvent, DatabaseLogEvent, DisruptionEvent, DbEventsFilter, SpotTerminationEvent,
+                             InfoEvent, CassandraStressEvent, ScyllaBenchEvent, CoreDumpEvent, DatabaseLogEvent, DisruptionEvent, DbEventsFilter, SpotTerminationEvent,
                              KillTestEvent, Severity, ThreadFailedEvent)
 
 LOGGER = logging.getLogger(__name__)
 
 logging.basicConfig(format="%(asctime)s - %(levelname)-8s - %(name)-10s: %(message)s", level=logging.DEBUG)
 
+# pylint: disable=too-many-public-methods
```
Comment (fruch) [inline unit_tests/test_events.py]: why here, and not on the calls it's related to ?

Comment (juliayakovlev) [reply unit_tests/test_events.py]: @fruch AFAIK it's per class. max-public-methods is 20 by default. May be we need to increase it?

File: test-cases/longevity/longevity-large-partition-4days.yaml [hunk 2]
```diff
@@ -1,14 +1,36 @@
...
+                     "scylla-bench -workload=sequential -mode=write -replication-factor=3 -partition-count=250 -clustering-row-count=100000 -partition-offset=251 -clustering-row-size=uniform:3072..5120 -concurrency=10 -connection-count=10 -rows-per-request=10 -timeout=30s",
+                     "scylla-bench -workload=sequential -mode=write -replication-factor=3 -partition-count=250 -clustering-row-count=100000 -partition-offset=501 -clustering-row-size=uniform:3072..5120 -concurrency=10 -connection-count=10 -rows-per-request=10 -timeout=30s",
+                     "scylla-bench -workload=sequential -mode=write -replication-factor=3 -partition-count=250 -clustering-row-count=100000 -partition-offset=751 -clustering-row-size=uniform:3072..5120 -concurrency=10 -connection-count=10 -rows-per-request=10 -timeout=30s"
                     ]
-n_db_nodes: 4
-n_loaders: 3
+prepare_verify_cmd: ["scylla-bench -workload=sequential -mode=read -replication-factor=3 -partition-count=250 -clustering-row-count=100000 -clustering-row-size=uniform:3072..5120 -concurrency=100 -connection-count=100 -rows-per-request=10 -timeout=30s -iterations 1 -validate-data",
```
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-4days.yaml]: The verify isn't splitted correctly, it should split only the 250 partitions and all should run with validate-data.
Also, worth splitting it to even more commands so each loader will run 2-3 commands to parallelize it even more (issue with scylla-bench).

File: unit_tests/test_events.py (hunk 1 above)
Comment (fruch) [reply unit_tests/test_events.py]: since this is a test, it's not that impotent, but my general comment here, that we should disable on a file level like that, this commend need to go in the same line of the class declaration it's related to. i.e. like that it's affecting the whole file, and other classes defined in this file.

if 20 is too low, maybe. but it's just a thumb rule, to help us think if a refactor is needed. that the whole point of those specific warnings. hitting them again and again and enlarging the number is like disabling them.

Comment (juliayakovlev) [reply unit_tests/test_events.py]: up it

**Comment by bentsi** [comment]
@juliayakovlev why it is a draft?

**Comment by juliayakovlev** [comment]
> @juliayakovlev why it is a draft?

missed that

**Comment by fruch** [comment]
@juliayakovlev seem like you have conflicts

@roydahan care to do a final review on this one ?

**Comment by juliayakovlev** [comment]
> @juliayakovlev seem like you have conflicts

I see all tests are passed

> 
> @roydahan care to do a final review on this one ?

**Comment by juliayakovlev** [comment]
> @juliayakovlev seem like you have conflicts

fixed

> 
> @roydahan care to do a final review on this one ?

**Comment by roydahan** [comment]
Yes, don't merge till I approve. in my queue.

**Comment by fruch** [comment]
@roydahan a reminder

@juliayakovlev seem like you have a small conflict

**Comment by roydahan** [comment]
AFAIK this isn't passed testing yet, right? (@juliayakovlev )
Or the issues you have are only with deletions nemesis?

File: sdcm/cluster.py [hunk 3]
```diff
@@ -3661,25 +3661,35 @@ def node_run_stress_bench(node, loader_idx, stress_cmd, node_list):
...
-                                                  stress_cmd, node_list))
+                                            args=args)
             setup_thread.daemon = True
             setup_thread.start()
             time.sleep(30)
 
+        if round_robin:
+            loader = self.get_loader()
+            loader_idx = [i for i, loader_obj in enumerate(self.nodes) if loader_obj == loader][0]
+            add_and_start_thread(args=(loader, loader_idx, stress_cmd, node_list))
```
Comment (roydahan) [inline sdcm/cluster.py]: Does this code works when you have more commands than loader?
less commands than loaders?
Not that readable to me.
Basically you could just have list of commands, num of loaders, each loader gets the MOD command that fits its index.

File: sdcm/cluster.py [hunk 4]
```diff
@@ -3661,25 +3661,35 @@ def node_run_stress_bench(node, loader_idx, stress_cmd, node_list):
...
+        else:
```
Comment (roydahan) [inline sdcm/cluster.py]: Then you won't need the else also

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                     "scylla-bench -workload=sequential -mode=read -replication-factor=3 -partition-count=15 -clustering-row-count=100000 -partition-offset=196 -clustering-row-size=uniform:3072..5120 -concurrency=100 -connection-count=100 -rows-per-request=10 -timeout=30s -iterations 1 -validate-data",
+                     "scylla-bench -workload=sequential -mode=read -replication-factor=3 -partition-count=15 -clustering-row-count=100000 -partition-offset=211 -clustering-row-size=uniform:3072..5120 -concurrency=100 -connection-count=100 -rows-per-request=10 -timeout=30s -iterations 1 -validate-data",
+                     "scylla-bench -workload=sequential -mode=read -replication-factor=3 -partition-count=25 -clustering-row-count=100000 -partition-offset=226 -clustering-row-size=uniform:3072..5120 -concurrency=100 -connection-count=100 -rows-per-request=10 -timeout=30s -iterations 1 -validate-data"
+                    ]
```
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: the count here is 25, not 15 as others.
Is it intentional?
Also, we should have a better code that generate these commands by itself and not manually calculate them.
You can do it in followup PR but this is not something we want to keep for long.

e.g. a template command that gets the number of parallelism and generate the list of commands.

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+             "scylla-bench -workload=sequential -mode=read -replication-factor=3 -partition-count=15 -clustering-row-count=100000 -partition-offset=151 -clustering-row-size=uniform:3072..5120 -concurrency=100 -connection-count=100 -rows-per-request=10 -timeout=30s -iterations 1 -validate-data",
+             "scylla-bench -workload=sequential -mode=read -replication-factor=3 -partition-count=42 -clustering-row-count=100000 -partition-offset=166 -clustering-row-size=uniform:3072..5120 -concurrency=100 -connection-count=100 -rows-per-request=10 -timeout=30s -iterations 1 -validate-data",
+             "scylla-bench -workload=sequential -mode=read -replication-factor=3 -partition-count=43 -clustering-row-count=100000 -partition-offset=208 -clustering-row-size=uniform:3072..5120 -concurrency=100 -connection-count=100 -rows-per-request=10 -timeout=30s -iterations 1 -validate-data"
+            ]
```
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: Same as above

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+# To validate rows in partitions: collect data about partitions and their rows amount
+# before and after running nemesis and compare it
+validate_partitions: true
+table_name: "scylla_bench.test"
+primary_key_column: "pk"
+
+# scylla-manager configuration
+# if running on aws and use_mgmt is true, the monitor image should not contain scylla
+use_mgmt: true
```
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: Why is it here? it should be default in test defaults.

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+round_robin: true
+
+instance_type_db: 'i3en.3xlarge'
+
+nemesis_class_name: 'ChaosMonkey'
+nemesis_interval: 30
+
+user_prefix: 'longevity-large-partitions-4d'
+
+space_node_threshold: 644245094
```
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: Do we really use it? if not remove, if yes what is the default and why this value?

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+n_db_nodes: 5
+n_loaders: 4
+n_monitor_nodes: 1
+
+round_robin: true
+
+instance_type_db: 'i3en.3xlarge'
+
+nemesis_class_name: 'ChaosMonkey'
```
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: Defaults?

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+nemesis_interval: 30
+
+user_prefix: 'longevity-large-partitions-4d'
```
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: I don't think it's needed anymore

File: vars/scyllaCloudLongevityPipeline.groovy [hunk 11]
```diff
@@ -62,24 +62,26 @@ def call(Map pipelineParams) {
             }
             stage('Create Cluster with siren-tests') {
                 steps {
-                        dir('siren-tests') {
-                            sh """
-                            #!/bin/bash
-                            set -xe
+                        wrap([$class: 'BuildUser']) {
```
Comment (roydahan) [inline vars/scyllaCloudLongevityPipeline.groovy]: How is this pipeline related to here?

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml (hunk 5 above)
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: > the count here is 25, not 15 as others.
> Is it intentional?

yes. I divided all rows by cmds amount

> Also, we should have a better code that generate these commands by itself and not manually calculate them.
> You can do it in followup PR but this is not something we want to keep for long.

Do you mean " not something we want to keep for long"? Is it temporary?

> 
> e.g. a template command that gets the number of parallelism and generate the list of commands.

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml (hunk 9 above)
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: Default is ```NoOpMonkey```

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml (hunk 7 above)
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: It's not in the defaults

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml (hunk 8 above)
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: It used when ```nemesis_during_prepare=true``` (that is default value). This value has come from original yaml. I think we need that. Need to wait for some prefill before start the nemesis.

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml (hunk 10 above)
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: Why not? All test has that

File: vars/scyllaCloudLongevityPipeline.groovy (hunk 11 above)
Comment (juliayakovlev) [reply vars/scyllaCloudLongevityPipeline.groovy]: It's not. I didn't touch it. Strange

File: sdcm/cluster.py (hunk 3 above)
Comment (juliayakovlev) [reply sdcm/cluster.py]: Longevity test send stress cmds one by one. To do it on the way that you suggest - it means to change longevity mechanism. We can do it as separate task.
Another solution - realize it on similat way as CassandraStressThread /GeminiThread

File: vars/scyllaCloudLongevityPipeline.groovy (hunk 11 above)
Comment (juliayakovlev) [reply vars/scyllaCloudLongevityPipeline.groovy]: It's out now

File: sdcm/cluster.py (hunk 3 above)
Comment (roydahan) [reply sdcm/cluster.py]: ok.

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml (hunk 5 above)
Comment (roydahan) [reply test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: I meant we can start with it and merge, but need to be improved to have "auto split".

File: test-cases/longevity/longevity-large-partition-200pk-4days.yaml (hunk 8 above)
Comment (roydahan) [reply test-cases/longevity/longevity-large-partition-200pk-4days.yaml]: ok

**Comment by roydahan** [comment]
safe to merge.

**Comment by juliayakovlev** [comment]
Jenkins job in master: https://jenkins.scylladb.com/view/master/job/scylla-master/job/longevity/job/longevity-large-partition-200k-pks-4days/

**Comment by bentsi** [comment]
we don't have branch-3.3 yet, @juliayakovlev did you mean `branch-3.2`?

**Comment by bentsi** [comment]
@juliayakovlev this is not easily back portable to 2019.1. Can you create a backport PR for this (if we really need this in 2019.1)?

**Comment by juliayakovlev** [comment]
> @juliayakovlev this is not easily back portable to 2019.1. Can you create a backport PR for this (if we really need this in 2019.1)?

@roydahan  Roy do we need to backport this pr to 2019.1? I think yes. Please, let your opinion

**Comment by roydahan** [comment]
Yes
